README.md

```diff
@@ -24,7 +24,7 @@ Specifies the messages exchanged between daemons during a swap and the peer-to-p
 ### [05. Tasks & Blockchain Events](./05-tasks-and-events.md)
 Specifies the tasks that syncers make available to daemons for the monitoring of chain-state through blockchain events and the broadcasting of transactions.
 
-### [06. Instructions & State Digests](./06-instructions-and-digests.md)
+### [06. Instructions](./06-instructions.md)
 Specifies the instructions that daemon makes available to clients for piloting the swap and state digests send to clients by the daemon.
```

Comment:
This needs to be updated

Reply:
State digest, as currently specified, is so abstract, and has no real connection to the instructions

``` md
    - [`u16`: `fired_transitions_len`]
    - [`fired_transitions_len * type`: `fired_transitions`] Vector of # of firings per transition since last state digest
    - [`sha256`: `marking_hash`] Hash of the current marking so that client can verify it applied the transitions correctly
    - [`u16`: `fireable_transition_data_len`]
    - [`fireable_transition_data_len * type`: `fireable_transition_data`] Map of fireable transitions to the data that the client requires to produce the authentication the daemon requires to fire it, such as transaction signatures
```

It does not even provide the information required for creating the instructions, such as

```
 transaction previously provided by the daemon via `state digest`
```

What i think belongs to instructions is the accessory data that must be included to the state-digest specification.

Reply:
You could think of it as the daemon provides the client the instructions it wants to receive back signed

These proposed instructions could be appended to the state digest

Reply:
@Lederstrumpf did u write the state-digest spec? can u participate in this discussion?

Reply:
yes.

> It does not even provide the information required for creating the instructions, such as
> "transaction previously provided by the daemon via `state digest`"

these transactions, like transaction signatures, are part of `fireable_transition_data`.

The idea is that since the client has its own petri net representation of the system state, it should also embed a mapping of the types of every transition's associated `transition_data` record.
So whenever the daemon provides a marking and the fireable transitions, the client parses the `fireable_transition_data` of each fireable transition using the type mapping already embedded in the client along with the petri net representation of the system state.

Reply:
Isn't it simpler for the daemon to directly encode into client's representation and talk to the client directly on client's own language?

Reply:
Yes, in fact the current design requires that the daemon encodes this in a way compatible with the client's representation (simplest is to encode it exactly as the client's representation consumes it, without any conversion on the client's part).

We could also do this more akin to TLV-spirit, i.e. to handle version discrepancies between client & daemon, i.e. nest TLV encoding within the `value` field (whenever it's a complex data structure) like `fireable_transition_data` too, but this is not explicitly specified in https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#type-length-value-format, so I went for the simpler route.

Reply:
Maybe we should not try to synchronize state explicitly, let the message itself be all the missing resources that will allow client to update its own state

Synchronizing state is hard. Daemon has its state, client has its own state, that is easier. Additionally the daemon doesn't control, the client does. The daemon shouldnt be telling the client what its state is, client should derive its state from past daemon messages and instruct daemon accordingly by making new resources available to daemon. That in turn signals the daemon to update its own state. 

Additionally, that fits better with the long term vision that eventually all components might talk to each other through message buses. That is, resources might be sourced directly.

Reply:
> Maybe we should not try to synchronize state explicitly, let the message itself contain all the missing resources that will allow client to update its own state

> Synchronizing state is hard. Daemon has its state, client has its own state, that is easier. Additionally the daemon doesn't control, the client does. The daemon shouldnt be telling the client what its state is, client should derive its state from past daemon messages and instruct daemon accordingly by making new resources available to daemon. That in turn signals the daemon to update its own state.

A few parallel points that bleed into one another:

1. If both client and daemon will use petri nets to represent the system state, then both of them having the same representation of the petri-net protocol is *simpler* than having to synchronize across distinct representations. We can do as you suggest here that the client "replays state transitions" to arrive at an internal representation of the protocol petri net's marking, but this requires the daemon to send information that is irrelevant to the client (only needed for replaying, not for creating signatures/instructions that the daemon can consume), so the daemon telling the client the shared protocol petri net's latest marking is the most efficient route.
2. In case I didn't mention this before: these state digests are restricted to the petri net representing the logic of the swap protocol, not any state of the daemon that the client does not have to care about, nor vice versa. The content of the state digest is limited to "what transitions of the protocol's petri net can currently be fired, and what do I (the daemon) require from you (the client) to fire this?". Thinking of this as an overall state synchronization makes this exchange of `fireable transitions + the information the client requires (to create a response that the daemon requires to fire this transition)` seem to have a much larger synchronization requirement than actually needed. And the protocol's state is the one essential aspect that both client & daemon should agree on, since a failure of this could be catastrophic: loss of funds.
3. There's a balancing act between loose coupling of components to allow them to evolve independently and also all talk to one another, against a principled design where the relationship between components is predictable. Maintaining a principled design incurs higher maintenance cost as software evolves & thus tends to scale worse, so I strongly prefer a loose coupling in most situations. In the context of daemon and client, having a shared model of the protocol state however is simpler than having separate models of the protocol state that we must translate in-between. So in this particular situation, a tight coupling of representations is desirable.
4. In fact, the daemon providing the client with a batch of "here's what I require you to sign to fire transition X" is a more loosely coupled communication model since the messages translate to "here's what I can do, and this is what I require to perform that service" versus "here's what I have - tell me what to do, and maybe I can". The former already tells the client what the available callbacks are - the latter assumes foreknowledge of these or a separate communication of them.
   
> Additionally, that fits better with the long term vision that eventually all components might talk to each other through message buses. That is, resources might be sourced directly.

The relationship between daemon and client is nonetheless the most sensitive & privileged since these components are fully trusted, so it's important that they agree on the protocol state - I don't see how this incompatible with them sharing this communication across a message bus shared by all components?

Additionally, the sharing of petri net states (or rather, the fireable transitions) does not preclude the direct sharing of resources, but it accompanies the resources (`fireable_transition_data`) to provide context for what the counterparty component will do with each of the possible responses it receives.

Reply:
ok, so let me translate what i read: u take a model for asynchronous distributed computation and make it synchronous, requiring synchronizing state (that is forbidden!). If async, daemon sends message m0 and then m1, and client receives them in wrong order, client is in the wrong state??? oh, to fix that we drop the async part and make it blocking. great!

Reply:
@Lederstrumpf if u r ok with this PR as of now, please let us know (u r the blocking factor)

I haven't modified anything to do with state_digest, the current blocking discussion, and will do so in a new, small PR

Reply:
> ok, so let me translate what i read: u take a model for asynchronous distributed computation and make it synchronous, requiring synchronizing state (that is forbidden!). If async, daemon sends message m0 and then m1, and client receives them in wrong order, client is in the wrong state??? oh, to fix that we drop the async part and make it blocking. great!

> u take a model for asynchronous distributed computation and make it synchronous, requiring synchronizing state (that is forbidden!)
How does this make petri-nets synchronous?

Wrt message ordering: nope, this does not lead to incorrect client states: since state digests from the daemon tell the client whatever the latest transitions available to be fired are, if the client receives `m1` and then `m0`, `m0` is irrelevant anyway because `m1` contains the latest state digest, so any delta between `m0` and `m1` would no longer be fireable transitions. If the client never receives `m1` but has received `m0` and instructs to fire a transition, the daemon will still fire it, if that transition is still valid for the latest state of the daemon.